File: PAT Advance/PAT_A1076.cpp
//https://github.com/QSCTech-Sange/PAT-in-Python/blob/master/1076-Forwards-on-Weibo.md
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

int main() {
    // 数据读入
    int num_users, max_level;
    scanf("%d%d", &num_users, &max_level);
    vector<int> fans[num_users + 1];
    for (int i = 1; i <= num_users; i++) {
        int num_friends;
        scanf("%d", &num_friends);
        for (int j = 0; j < num_friends; j++) {
            int friend_temp;
            scanf("%d", &friend_temp);
            fans[friend_temp].push_back(i);
        }
    }
    // 对每个输入进行BFS
    int num_que;
    scanf("%d", &num_que);
    for (int i = 0; i < num_que; i++) {
        int id;
        scanf("%d", &id);
        bool visited[num_users + 1];
        for (int j = 0; j <= num_users; j++)
            visited[j] = false;
        visited[id] = true;
        int count = 0;
        int level = 1;
        queue<int> que;
        que.push(id);
        while (!que.empty() and level <= max_level) {
            int length = que.size();
            for (int j = 0; j < length; j++) {
                int star = que.front();
                que.pop();
                for (auto &fan:fans[star]) {
                    if (!visited[fan]) {
                        visited[fan] = true;
                        que.push(fan);
                        count += 1;
                    }
                }
            }
            level += 1;
        }
        printf("%d\n", count);
    }
}

File: PAT Advance/PAT_A1076.py
'''
参考六度分隔
广度优先搜索，增加层号记录
'''

N, L = map(int,input().split())

LGraph = [ [] for i in range(N+1)]

for item in range(1,N+1):
    inputLine = list(map(int,input().split()))
    if inputLine[0] != 0:
        for user in inputLine[1::]:
            LGraph[user].append(item)

UserID = list(map(int,input().split()))

# N,L = 7,3
# LGraph = [[], [4], [1], [1, 4, 5], [1, 5, 6], [3, 7], [3], []]
# UserID = [2, 2, 6]


def BFS(vertex):
    visited = [0]*(N+1)
    Queue = []
    visited[vertex] = True
    Queue.append(vertex)
    count = 0
    level = 0
    tail = 0
    last = vertex
    while Queue and level<L:
        vQ = Queue.pop(0)
        for item in LGraph[vQ]:
            if visited[item] != True:
                Queue.append(item)
                visited[item] = True
                count += 1
                tail = item
        
        if vQ == last:
            level+=1
            last = tail

        if level == 6:
            break

    return count


for user in UserID[1::]:
    print(BFS(user))